From 1e47554dc394cba458732bb5c5278a22e3ac36ed Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Wed, 14 Oct 2026 09:44:20 +0800
Subject: [PATCH 63/68] add hack for old-world seccomp sandbox

Add back __NR_fstat and __NR_newfstatat so glibc will use fstat and
newfstatat instead of statx for anything other than statx itself, see
commit 551101e8240b ("io: Do not implement fstat with fstatat").  Also
undefine __ASSUME_STATX to allow statx() itself to fall back in case
some (stupid I'd say) software calls statx() but blacklists the statx
syscall.

Note that this issue is not limited to ABI1.0 (i.e. the fstat/fstatat
implementation is orthogonal to ABI1.0/ABI2.0) but if you hit this issue
on ABI2.0 you should collaborate with the upstreams to figure out an
elegant solution (based on the nullable path support with AT_EMPTY_PATH
on Linux >= 6.11 for statx) instead of using this hack or more stupid
hack (for example, forcing a lower --enable-kernel-version which is done
by some commercial or so-called "community" distros and makes a
Frankenstein ABI).

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 .../sysv/linux/loongarch/fixup-asm-unistd.h   | 21 -------------------
 .../sysv/linux/loongarch/lp64/arch-syscall.h  |  2 ++
 .../linux/loongarch/lp64/kernel-features.h    |  6 ++++++
 3 files changed, 8 insertions(+), 21 deletions(-)
 delete mode 100644 sysdeps/unix/sysv/linux/loongarch/fixup-asm-unistd.h
 create mode 100644 sysdeps/unix/sysv/linux/loongarch/lp64/kernel-features.h

diff --git a/sysdeps/unix/sysv/linux/loongarch/fixup-asm-unistd.h b/sysdeps/unix/sysv/linux/loongarch/fixup-asm-unistd.h
deleted file mode 100644
index a823a26ad2..0000000000
--- a/sysdeps/unix/sysv/linux/loongarch/fixup-asm-unistd.h
+++ /dev/null
@@ -1,21 +0,0 @@
-/* Regularize <asm/unistd.h> definitions.  LoongArch version.
-   Copyright (C) 2024-2026 Free Software Foundation, Inc.
-
-   The GNU C Library is free software; you can redistribute it and/or
-   modify it under the terms of the GNU Lesser General Public
-   License as published by the Free Software Foundation; either
-   version 2.1 of the License, or (at your option) any later version.
-
-   The GNU C Library is distributed in the hope that it will be useful,
-   but WITHOUT ANY WARRANTY; without even the implied warranty of
-   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-   Lesser General Public License for more details.
-
-   You should have received a copy of the GNU Lesser General Public
-   License along with the GNU C Library; if not, see
-   <http://www.gnu.org/licenses/>.  */
-
-/* To avoid the messy usage of the fstat, newfstatat, and statx system calls, we
-only use statx.  */
-#undef __NR_fstat
-#undef __NR_newfstatat
diff --git a/sysdeps/unix/sysv/linux/loongarch/lp64/arch-syscall.h b/sysdeps/unix/sysv/linux/loongarch/lp64/arch-syscall.h
index f7294acaad..121a148803 100644
--- a/sysdeps/unix/sysv/linux/loongarch/lp64/arch-syscall.h
+++ b/sysdeps/unix/sysv/linux/loongarch/lp64/arch-syscall.h
@@ -61,6 +61,7 @@
 #define __NR_fsmount 432
 #define __NR_fsopen 430
 #define __NR_fspick 433
+#define __NR_fstat 80
 #define __NR_fstatfs 44
 #define __NR_fsync 82
 #define __NR_ftruncate 46
@@ -172,6 +173,7 @@
 #define __NR_munmap 215
 #define __NR_name_to_handle_at 264
 #define __NR_nanosleep 101
+#define __NR_newfstatat 79
 #define __NR_nfsservctl 42
 #define __NR_open_by_handle_at 265
 #define __NR_open_tree 428
diff --git a/sysdeps/unix/sysv/linux/loongarch/lp64/kernel-features.h b/sysdeps/unix/sysv/linux/loongarch/lp64/kernel-features.h
new file mode 100644
index 0000000000..a4c343cf31
--- /dev/null
+++ b/sysdeps/unix/sysv/linux/loongarch/lp64/kernel-features.h
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include_next <kernel-features.h>
+
+/* Allow falling back to syscalls "easier to audit" for statx() itself.  */
+#undef __ASSUME_STATX
-- 
2.56.0

